pkg/k8smgmt/ingress.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2024 EdgeXR, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8smgmt
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/edgexr/edge-cloud-platform/api/distributed_match_engine"
+	"github.com/edgexr/edge-cloud-platform/api/edgeproto"
+	"github.com/edgexr/edge-cloud-platform/pkg/log"
+	"github.com/edgexr/edge-cloud-platform/pkg/platform/pc"
+	ssh "github.com/edgexr/golang-ssh"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+)
+
+const (
+	IngressClassName = "nginx"
+)
+
+// CreateIngress creates an ingress to handle HTTP ports for the
+// AppInst. We assume each AppInst has its own host name,
+// and an AppInst does not need more than one host name.
+// The AppInst host name must match the wildcard cert for the
+// cluster/cloudlet.
+// TODO: This assumes the AppInst is in a single namespace and
+// creates a single ingress.
+// For complex AppInsts (helm charts) it may need create an
+// ingress per namespace if the AppInst uses multiple namespaces.
+func CreateIngress(ctx context.Context, client ssh.Client, names *KubeNames, appInst *edgeproto.AppInst) (*networkingv1.Ingress, error) {
+	log.SpanLog(ctx, log.DebugLevelInfra, "creating ingress", "appInst", appInst.Key.GetKeyString())
+
+	ingressClass := IngressClassName
+
+	ingress := networkingv1.Ingress{}
```

Comment:
We generally use templates to generate a k8s manifest. What was the reason to use the k8s objects here?

Reply:
I thought it would be easier to deal with the port logic that way. Although I could have also generated an intermediate "args" object after applying the logic and then used that with the template. I think a template would have worked fine as well.

One thing I was thinking about though, is that we may want to move to using the [go-client](https://github.com/kubernetes/client-go) instead of using kubectl. We already reference that library in our code, and it solves one pain point I was dealing with with the CCRM, where we always need to rewrite the kubeconfig on every operation (or read it and make sure it's correct) to deal with CCRM restarts, or commands being run on different CCRMs. If we use the client, we don't need to write a kubeconfig to disk (we just load it in memory in the client), so that issue goes away. Also any issues with conflicting kubeconfigs for different apps/clusters/etc on disk goes away. Also it's more efficient than spawning a separate process (kubectl). But it's a big change from how we do things currently, so not something I want to take on now. Also there's one big issue with the go-client, that we can no longer go via an intermediate node via ssh, which would break the current VM-based plataforms.